### PR TITLE
Fixed #25095 -- Fixed annotate() + values() group by bug

### PR DESCRIPTION
**Pull request by mitchelljkotler**

### CODE REVIEW

**Comment by timgraham** [comment]
buildbot, add to whitelist.

File: tests/aggregation/tests.py [hunk 1]
```diff
@@ -395,6 +395,25 @@ def test_count(self):
         vals = Book.objects.aggregate(Count("rating", distinct=True))
         self.assertEqual(vals, {"rating__count": 4})
 
+    def test_annotate_values_count(self):
+        vals = Book.objects.annotate(xprice=F('price')).filter(xprice__lte=30).values('publisher').distinct().annotate(count=Count('pk')).order_by('publisher')
```
Comment (jarshwah) [inline tests/aggregation/tests.py]: Please break this down into a couple of lines to make it easier to read. Also, the `distinct` call should be unnecessary for this bug, and only introduces extra work that distracts from the main problem.

File: tests/aggregation/tests.py [hunk 2]
```diff
@@ -395,6 +395,25 @@ def test_count(self):
         vals = Book.objects.aggregate(Count("rating", distinct=True))
         self.assertEqual(vals, {"rating__count": 4})
 
+    def test_annotate_values_count(self):
```
Comment (jarshwah) [inline tests/aggregation/tests.py]: I'd prefer a test name like `test_nonselected_annotations_not_in_groupby`, to clarify the purpose of the test.

**Comment by timgraham** [comment]
Mitchell, did you see the reply on the ticket about the additional test that's needed? Please uncheck "Patch needs improvement" on the ticket when you update the PR. Thanks.

File: tests/aggregation/tests.py [hunk 3]
```diff
@@ -395,6 +395,57 @@ def test_count(self):
...
+                },
+                {
+                    'publisher': 3,
+                    'count': 1,
+                },
+            ]
+        )
+
+    def test_nonselected_annotations_not_in_groupby2(self):
+        vals = list(Book.objects.annotate(xprice=F('price'))
```
Comment (timgraham) [inline tests/aggregation/tests.py]: A style that didn't require so many lines and use non-4-space indent would be more consistent with the rest of the code. Something like:
``` 
vals = list(Book.objects.annotate(xprice=F('price')).filter(xprice__lte=30)
    .values('publisher', 'contact').annotate(count=Count('pk')).values('publisher', 'count')
    .order_by('publisher'))
```

File: tests/aggregation/tests.py [hunk 4]
```diff
@@ -395,6 +395,57 @@ def test_count(self):
...
+                                .filter(xprice__lte=30)
+                                .values('publisher', 'contact')
+                                .annotate(count=Count('pk'))
+                                .values('publisher', 'count')
+                                .order_by('publisher'))
+        self.assertEqual(
+            vals, [
+                {
```
Comment (timgraham) [inline tests/aggregation/tests.py]: It would be fine to put these on a single line `{'publisher': 1, 'count': 1}`

**Comment by timgraham** [comment]
You may squash commits and use the commit message from the pull request title when updating, thanks!

File: tests/aggregation/tests.py [hunk 5]
```diff
@@ -395,6 +395,29 @@ def test_count(self):
...
+                        .annotate(count=Count('pk')).order_by('publisher'))
+        self.assertEqual(
+            vals, [
+                {'publisher': 1, 'count': 2},
+                {'publisher': 2, 'count': 1},
+                {'publisher': 3, 'count': 1},
+            ]
+        )
+
+    def test_nonselected_annotations_not_in_groupby2(self):
```
Comment (jarshwah) [inline tests/aggregation/tests.py]: I think the name of this test should be more clear - to signal the intent of the test. Maybe "test_grouped_annotation_in_groupby"?

**Comment by jarshwah** [comment]
Other than the name of the second test, I think this is about good to go. Thanks @mitchelljkotler for owning this.

I still have a small concern about the tests themselves though. We're trying to ensure that annotations are added to the group by, or not, depending on the context of the query - namely the position and content of `.values()`. But we're not actually testing that the right columns are *in* the group by. 

For example, both of the below queries return the same results:

```
Book.objects.annotate(xprice=F('price')).values('publisher', 'xprice')
    .annotate(count=Count('pk')).values('publisher', 'count')

Book.objects.annotate(xprice=F('price')).values('publisher', 'pk')
    .annotate(count=Count('pk')).values('publisher', 'count')
```

Should we make this test (test_nonselected_annotations_not_in_groupby2) check the actual string content of the query being passed? I'm usually reluctant to do these kinds of tests in "frontend" modules, since different backends (no-sql?) probably generate different strings.

I think a better solution is to ensure that the query is constructed in such a way that it will definitely fail if any part of the query is modified or the backend changes. It'll force someone to look at the failing test and either fix the regression, or fix the test if results should be different.

And after some trial and error, I've come up with the following tests which differ slightly from yours:

```python
def test_nongrouped_annotation_not_in_groupby(self):
        """
        Tests that an annotation not included in `values()`, before an aggregate, is
        excluded from the group by clause.
        """
        qs = (
            Book.objects.annotate(xprice=F('price')).filter(rating=4.0).values('rating')
                .annotate(count=Count('publisher_id', distinct=True)).values('count', 'rating').order_by('rating')
        )
        self.assertEqual(
            list(qs), [
                {'rating': 4.0, 'count': 2},
            ]
        )

    def test_grouped_annotation_in_groupby(self):
        """
        Tests that an annotation included in `values()`, before an aggregate, is
        included in the group by clause.
        """
        qs = (
            Book.objects.annotate(xprice=F('price')).filter(rating=4.0).values('rating', 'xprice')
                .annotate(count=Count('publisher_id', distinct=True)).values('count', 'rating').order_by('rating')
        )
        self.assertEqual(
            list(qs), [
                {'rating': 4.0, 'count': 2},
                {'rating': 4.0, 'count': 1},
            ]
        )
```

These two tests, side by side, make it really clear that the only difference between each query is the addition of the `xprice` parameter. The values of these objects are nicely aligned to show the aggregation differs in each case. Are you OK with changing the tests to the ones I've written above @mitchelljkotler? I know there's been some back and forth, but that's it for my suggested edits :smile:

**Comment by manfre** [comment]
"different backends (no-sql?) probably generate different strings."

Mssql would also generate a different SQL string. 

The new tests look good to me.

**Comment by mitchelljkotler** [comment]
Thanks for the feedback. This looks good to me.  Unfortunately I'm away and
won't be able to make these updates for a couple weeks now.  If you can put
these in yourself, that be great, otherwise I will update it when I am back.

**Comment by timgraham** [comment]
Updated in https://github.com/django/django/pull/5041, but the second test doesn't pass on Oracle because the results are flipped (`order_by('rating')` is there, but the ratings are the same). Is it okay to order by count instead?

**Pull request closed without merging**